input/amqp.go

```diff
@@ -144,12 +144,62 @@ type amqpConnector func(a *Amqp) error
 // AMQPConnector connects using the given configuration
 func AMQPConnector(a *Amqp) error {
 	log.Infof("dialing AMQP: %v", a.uri)
-	conn, err := amqp.Dial(a.uri.String())
-	if err != nil {
-		return err
+
+	var conn *amqp.Connection
+	var err error
+
+	for {
+		config := amqp.Config{
+			Heartbeat: time.Duration(a.config.Amqp.Amqp_heartbeat) * time.Second,
+		}
+
+		conn, err = amqp.DialConfig(a.uri.String(), config)
+		if err == nil {
+			break
+		}
+
+		if !a.config.Amqp.Amqp_retry {
+			return err
+		}
+
+		log.Errorf("Failed to connect to AMQP server: %v. Retrying in %d seconds...", err, a.config.Amqp.Amqp_retrydelay)
+		time.Sleep(time.Duration(a.config.Amqp.Amqp_retrydelay) * time.Second)
 	}
+
 	a.conn = conn
 
+	// Create a channel to receive close notifications from the connection
+	closeCh := make(chan *amqp.Error)
+	conn.NotifyClose(closeCh)
+
+	// Start a goroutine to monitor the connection state
+	go func() {
+		for {
+			select {
+			case <-closeCh:
+				log.Println("AMQP connection closed.")
+
+				if !a.config.Amqp.Amqp_retry {
+					log.Println("Retry is disabled. Exiting reconnection attempt.")
+					return
+				}
+
+				log.Println("Attempting to reconnect...")
+
+				for {
+					err := AMQPConnector(a)
+					if err == nil {
+						log.Println("Successfully reconnected to AMQP server.")
+						return
+					}
+
+					log.Errorf("Failed to reconnect to AMQP server: %v. Retrying in %d seconds...", err, a.config.Amqp.Amqp_retrydelay)
+					time.Sleep(time.Duration(a.config.Amqp.Amqp_retrydelay) * time.Second)
+				}
```

Comment:
Is this `for`-loop needed here? The recursive call to `AMQPConnector` will take care of retrying, and will also register the `closeCh` to the new connection. In fact, the recursive call here will never return an error if the connection fails, it will retry (if retry is enabled). 

If I'm following this correctly, the `for` should be replaces with a single call to `AMQPConnector(a)`

Reply:
yep, I got your point, retry/retrydelay are already built-in on `AMQPConnector(a)`, so the loop here seems to be redundant/duplicated code... I will test it, and I will come back with an amend when I validate that works as expected, nice catch!

Reply:
Thank you again for contributing! LMK and I'll take another look

Reply:
@npazosmendez, the fix has been added. The code now looks much better and cleaner. I have built and tested it, and everything seems to be working well. Let me know what you think. Thanks for your comments and review; I really appreciate it!
```
carbon-relay-ng1    | 2024-06-13 20:09:09.573 [INFO] Successfully connected to AMQP server: amqp://admin:password@192.168.248.146/.
carbon-relay-ng1    | 2024-06-13 20:09:09.773 [INFO] consuming AMQP messages
carbon-relay-ng2    | 2024-06-13 20:09:09.153 [INFO] listening on 0.0.0.0:2003/udp
carbon-relay-ng2    | 2024-06-13 20:09:09.153 [INFO] dialing AMQP: amqp://admin:password@192.168.248.146/
carbon-relay-ng2    | 2024-06-13 20:09:09.154 [INFO] listening on 0.0.0.0:2013/tcp
carbon-relay-ng2    | 2024-06-13 20:09:09.154 [INFO] listening on 0.0.0.0:2003/tcp
carbon-relay-ng2    | 2024-06-13 20:09:09.154 [INFO] admin HTTP listener starting on 0.0.0.0:8081
carbon-relay-ng2    | 2024-06-13 20:09:09.156 [INFO] dest carbon-default_192_168_242_31_2003 new conn online
carbon-relay-ng2    | 2024-06-13 20:09:09.156 [INFO] dest carbon-default_192_168_242_33_2003 new conn online
carbon-relay-ng2    | 2024-06-13 20:09:09.157 [INFO] dest carbon-default_192_168_242_32_2003 new conn online
carbon-relay-ng2    | 2024-06-13 20:09:09.356 [INFO] Successfully connected to AMQP server: amqp://admin:password@192.168.248.146/.
carbon-relay-ng2    | 2024-06-13 20:09:09.556 [INFO] consuming AMQP messages
carbon-relay-ng3    | 2024-06-13 20:09:09.377 [INFO] admin TCP listener starting on 0.0.0.0:2004
carbon-relay-ng3    | 2024-06-13 20:09:09.378 [INFO] listening on 0.0.0.0:2013/udp
carbon-relay-ng3    | 2024-06-13 20:09:09.377 [INFO] listening on 0.0.0.0:2003/tcp
carbon-relay-ng3    | 2024-06-13 20:09:09.378 [INFO] dialing AMQP: amqp://admin:password@192.168.248.146/
carbon-relay-ng3    | 2024-06-13 20:09:09.379 [INFO] admin HTTP listener starting on 0.0.0.0:8081
carbon-relay-ng3    | 2024-06-13 20:09:09.380 [INFO] dest carbon-default_192_168_242_33_2003 new conn online
carbon-relay-ng3    | 2024-06-13 20:09:09.380 [INFO] dest carbon-default_192_168_242_31_2003 new conn online
carbon-relay-ng3    | 2024-06-13 20:09:09.380 [INFO] dest carbon-default_192_168_242_32_2003 new conn online
carbon-relay-ng3    | 2024-06-13 20:09:09.576 [INFO] Successfully connected to AMQP server: amqp://admin:password@192.168.248.146/.
carbon-relay-ng3    | 2024-06-13 20:09:09.775 [INFO] consuming AMQP messages
carbon-relay-ng2    | 2024-06-13 20:10:01.002 [INFO] stats now connected to 192.168.242.33:2003
carbon-relay-ng3    | 2024-06-13 20:10:01.002 [INFO] stats now connected to 192.168.242.33:2003
carbon-relay-ng1    | 2024-06-13 20:10:01.002 [INFO] stats now connected to 192.168.242.33:2003
carbon-relay-ng3    | 2024-06-13 20:10:05.350 [INFO] AMQP connection closed.
carbon-relay-ng3    | 2024-06-13 20:10:05.350 [INFO] Attempting to reconnect...
carbon-relay-ng3    | 2024-06-13 20:10:05.350 [INFO] dialing AMQP: amqp://admin:password@192.168.248.146/
carbon-relay-ng1    | 2024-06-13 20:10:05.350 [INFO] AMQP connection closed.
carbon-relay-ng1    | 2024-06-13 20:10:05.350 [INFO] Attempting to reconnect...
carbon-relay-ng1    | 2024-06-13 20:10:05.350 [INFO] dialing AMQP: amqp://admin:password@192.168.248.146/
carbon-relay-ng2    | 2024-06-13 20:10:05.350 [INFO] AMQP connection closed.
carbon-relay-ng2    | 2024-06-13 20:10:05.351 [INFO] Attempting to reconnect...
carbon-relay-ng2    | 2024-06-13 20:10:05.351 [INFO] dialing AMQP: amqp://admin:password@192.168.248.146/
carbon-relay-ng3    | 2024-06-13 20:10:05.398 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng2    | 2024-06-13 20:10:05.399 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng1    | 2024-06-13 20:10:05.399 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng1    | 2024-06-13 20:10:35.449 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng2    | 2024-06-13 20:10:35.449 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng3    | 2024-06-13 20:10:35.449 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng1    | 2024-06-13 20:11:05.500 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng3    | 2024-06-13 20:11:05.500 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng2    | 2024-06-13 20:11:05.500 [ERROR] Failed to connect to AMQP server: dial tcp 192.168.248.146:5672: connect: connection refused. Retrying in 30 seconds...
carbon-relay-ng2    | 2024-06-13 20:11:35.705 [INFO] Successfully connected to AMQP server: amqp://admin:password@192.168.248.146/.
carbon-relay-ng1    | 2024-06-13 20:11:35.705 [INFO] Successfully connected to AMQP server: amqp://admin:password@192.168.248.146/.
carbon-relay-ng3    | 2024-06-13 20:11:35.705 [INFO] Successfully connected to AMQP server: amqp://admin:password@192.168.248.146/.
```